### ISSUE TITLE
null casting inconsistent behaviour

### ISSUE THREAD

**Issue by yanhick** [HaxeFoundation/haxe#3848]
There seems to be different semantic for casting null in haxe cpp. The following will throw an error in Js and Neko but not in Cpp
```
cast(null, MyClass);
```
Note that if I cast primitive types (String, Int...), Js, Neko and Cpp all throw an error. I didn't try other targets. I'm using Haxe 3.1.2.

This could be fixed in 2 ways:
- Allow null casting on all types. Java does that.
- Prevent casting null to an object in haxe cpp.

What do you think ?

**Comment by ncannasse** [comment]
I agree that `cast(null,T)` should return null, since it's a valid value. We should indeed unit test and fix this accordingly, thanks for reporting.

**Comment by bji** [comment]
Are we also covering the runtime case here?  i.e.

var foo = someFunctionThatReturnsNull();

var myClass = cast(foo, MyClass);

Should this also silently allow the runtime value null to be cast to type MyClass?

**Comment by Simn** [comment]
I don't think I like this. What value does silently allowing null have here? Almost anything you try to do with the value after the cast succeeds is going to error anyway. In the few cases it doesn't it's probably not going to behave as intended either and might be quite difficult to track down.

**Comment by bji** [comment]
You can do "if (myClass == null)", or assign some other variable's value to myClass.  The point being that null is a valid value for an instance of a myClass, and you can do all the stuff you would normally do with such a null reference.

**Comment by Simn** [comment]
Sure but you might as well do that null check before trying the cast. At least then the special case is separated instead of being nested in a cast case. That just leads to weird code in my opinion.

You also create new problems with regards to basic types. What if I try to safe-cast null to Int on a static target?

**Comment by waneck** [comment]
Unless we detect cast `(x, Null <X>)`, I'm not sure that disallowing null
is good. Also this would need an extra check for Java/C#, which is kind of
annoying

**Comment by bji** [comment]
It's true that you might as well do the null check before the cast, but that's not a reason to disallow doing it after.  Also, waneck, are you supporting allowing casting from a null runtime value or supporting disallowing it?  Your comment isn't clear to me.  Thanks!

**Comment by Simn** [comment]
Do you also expect Std.is to return true if the first argument is null (run-time or not)?

It is currently documented to return false if v is null. We would end up with a really weird API if we keep that and go for "null is a valid value for all types" for safe-casts.

**Comment by bji** [comment]
Yes, I would expect Std.is to return true for any null value evaluated at runtime.  I would also expect a compile time test that would replace the Std.is with a constant true or false value if the Std.is can be calculated at compile time.

**Comment by ncannasse** [comment]
I don't think we should change Std.is behavior. Std.is test the "runtime" type of a value, and null has no runtime type associated. cast is another story, since it checks if the runtime value is compatible with the type, which null is. At least that's how I would like them to be specified.

**Comment by Simn** [comment]
The distinction doesn't make any sense to me. We even implement safe-casting as `if (Std.is(v, T)) cast v else throw "Class cast error"` on several targets.

I think if we change one we should change the other, and we definitely cannot change Std.is at the moment so I'm in favor of not changing anything for the time being.

**Comment by ncannasse** [comment]
Not changing something that was reported as a platform inconsistency is just making it worse, if we consider that the amount of Haxe code written will grow over the years.

**Comment by nadako** [comment]
am i right thinking that you want to (e.g. on C#) generate `(o == null) ? null : (SomeType)o` instead of simple `(SomeType)o`? This is quite horrible for performance. Also `null` is not a valid value for basic types on static targets, tho we still allow `cast(v,Int)`, what should we do in this case? I think we should throw exception on safe-casting `null`.

**Comment by bji** [comment]
I agree with ncannasse's most recent statement.  The integrity of the language going forward is more important than backwards compatibility with incorrect prior implementations, and I think that major language versions is a good time to make such changes, so I would recommend that if this is changed, it should be done on the next haxe major version (4.0), and well documented.  I have a feeling that there is very little code out there that depends upon the specific behavior of detecting casts from null via catching exceptions instead of just direct testing of the variable for null, so preventing that exception and allowing a cast from null is unlikely to break much existing code.

In terms of Std.is, it is merely an optimization to make a compile time check and to turn the expression into a constant true or false in some situations.  At runtime, how can something like:

var foo = new MyClass(); // MyClass has no super class

if (Std.is(foo, String)) ...

ever evaluate to anything other than false in the Std.is call?  foo will never be an instance of String and the compiler can know this because it knows the type of foo at compile time and knows that the Std.is call will always return false in its runtime evaluation, so why not have the compiler optimize out the if statement?

Re: nadako's most recent post, I don't know the semantics of C# cast but assuming that it allows casting from null just like Java does, then I think the alternatives would look something more like this:

my position:

(SomeType) o

Simn's position:

if (o == null) throw new BadCastException() else return (SomeType) o

All that being said, I think that it's a big stretch to call Simn's position "quite horrible for performance".  We're talking about a single conditional check here which is going to be completely lost in the noise in the vast majority of use cases, and furthermore, in the places where it would even be noticeable, i.e. inner loops, just don't do casts there.  Do your casts outside the inner loop.

But of course, I still prefer my position, which I believe results in the least runtime code for most platforms and also provides the most consistent language semantics.

**Comment by waneck** [comment]
@bji, I prefer that we allow nulls when casting. This will play nicely with the Java/C# targets (as @nadako has mentioned), and afaik we don't allow `cast(x,Null<Something>)`, so it's not even possible to make a safe cast that would allow `Null<>`

**Comment by Simn** [comment]
I'm starting to have a hard time following who agrees with whom here. My position is that we can (and should) change it for Haxe 4, but shouldn't make a half-assed change before that. Sure it's nice to fix platform-inconsistencies early, but if that comes at the price of an API inconsistency (safe casts being different than Std.is) I really don't think it's worth it.

**Comment by nadako** [comment]
I just tested C# and I was wrong (SomeType)null works on c# as long as SomeType is reference type, so the `cast(v,Int)` question is still open.

**Comment by nadako** [comment]
If we're talking about haxe 4, can we please add proper null safety? :)

**Comment by bji** [comment]
Thanks for your clarification Simn.  Your position is actually quite different than I thought it was - I thought you were arguing against changing anything related to null casting and Std.is, even in a major release.

To clarify my position: I think we should allow casting from null, because it makes semantic sense to me (null is a valid value for any object reference type, so it should be castable to any object reference type), and also, it is consistent with the other languages I am aware of (Java, C++, and if I understand nadako correctly, also C#), and as a result, it results in the least generated code for most/all targets.

The remaining issue is about the behavior of Std.is(a, b): My position is that Std.is should return true for all a, b such that a cast from a to b will succeed at runtime without throwing an exception, and should return false otherwise.  This means that Std.is(null, AnyObjectTypeOrValue) will always return true.

**Comment by bji** [comment]
Also, I agree (as I stated earlier) that these changes should be done in a major release, not in a point release.

**Comment by Simn** [comment]
Related: #2976

**Comment by sganapavarapu1** [comment]
I don't think we should change Std.is behavior. 

There's code out there that's reliant on Std.is returning false when object is null.
e.g.
if (Std.is(obj, MyType))  {
  var temp : MyClass = cast(obj, MyType).getMyClass();
}
This will result in NPE when obj is null if Std.is returns true.

One may comment that, it is easy to fix above code to:
if (Std.is(o, MyType)) {
  var temp : MyClass;
  if (obj != null)  {
      temp = cast(obj, MyType).getMyClass();
  }
  else   {
       // null handling, if desired
  }
}

There are two things to consider when making such comment:
a) Volume of code that's going to be impacted --- however simple the fix may seem... 
b) Variants that can't be addressed by simple example quoted here. I can definitely pull few more variants that are not as straight forward as above. They are not complex but, same cookie cutter approach cannot be used for all.

Having said that all that, if I block myself from the thinking about the impact of changing Std.is... 
I like the idea of Std.is being in harmony with proposed fix for null handling in cast. 

But my final vote is against changing Std.is behavior as it breaks a lot of functionality across million of lines of code that we have.

**Comment by Simn** [comment]
We might want to review safe casts on a larger scale for Haxe 4. It often causes unexpected performance issues because people do not realize that it involves a run-time type check. Your first snippet is an example for that: depending on the target you're essentially doing a double-Std.is because the safe cast has one built-in.

**Comment by ncannasse** [comment]
I don't think too that we need to change Std.is, it's been specified and tested in a cross platform manner. We could change its specification in case we replace it by a "is" operator in Haxe 4.0, but that's another feature request.

The current report is about cross platform inconsistencies of safe-cast operator, which needs and should be addressed in Haxe 3.2.

If T is a nullable type, then cast(null,T) should return null instead of throwing an exception on all platforms. I don't think people are relying anyway on a particular exception to be thrown (since it's already platform specific), and they should not anyway.

If T is a basic type (Int/Float/Bool/etc), then it should return the platform default value (either null for non static platforms or 0/NaN depending on platform default.

**Comment by Simn** [comment]
I don't understand the rush here. It has been like this for years.

At least right now all the type-checking is consistent _within a target_. Introducing an inconsistency like this between safe-casts and Std.is is something PHP would do in an ad hoc fashion. Just imagine the conversation:

User: Hey compiler, can I cast this value to type T?
Haxe: Absolutely!
User: So it is an instance of type T?
Haxe: Nope!

I really don't understand why you consider these concepts detached to the point where you'd accept them giving opposite answers to the same questions.

**Comment by ncannasse** [comment]
@Simn I don't see any issue here, that seems somewhat to match the definition of null.

**Comment by bji** [comment]
Saying that it's been like this for years is, no offense, not at all relevant to the question of how it should be going forward.  Yes, this behavior, similarly for many behaviors, has been different for different targets for a while now, and I personally find this to be a shortcoming of haxe: it promises me that I can write code that will work well on lots of platforms, but there are many rough edges were that promise is not delivered and the runtime fails in different and weird ways depending upon the target.  The issue reported here is such an example.  But the good news is, that we can improve this!

Do you actually believe that it's better for the language to have different semantics for fundamental language operations like type casting depending upon the target?  If so, then I think we have a larger philosophical gap than is likely to be bridged in this discussion.

If you are unwilling to consider changing the semantics of Std.is to match the semantics of cast, then perhaps you'd consider just adding a new function to Std, something like Std.canCast, that has the semantics that I am describing.  Ideally this would instead be the 'as' expression that ncannasse has referred to, but if that's unachievable, then Std.canCast would be a reasonable alternative.  So your example would then be:

User: Hey compiler, can I cast this value to type T?
Haxe: Absolutely!
(user goes away and casts it and is happy)
User: But compiler, Std.is this a type T?
Haxe: Nope!  It Std.isn't!  But Std.is doesn't mean what you thought it did, which is why you stopped using it and started using Std.canCast instead.  So we're good, right?

**Comment by Simn** [comment]
And to me the "doesn't mean what you thought it did" is exactly the problem. It's an API which has surprises and I find that surprise-by-design is much worse than surprise-by-issue-not-being-addressed-yet.

But I give up; you guys seem to agree on something and I don't want to get in the way with my concerns which may or may not be rooted in me not understanding what `null` means.

**Comment by bji** [comment]
Well I don't really want you to give up, I'd rather that we could come to an agreement.  I think that having one of the primary language maintainers buying into the semantics surrounding casting is kind of important.

What do you mean when you say "not understanding what null means"?

Also, can you describe what you mean as surprise-by-design in the proposal?  Are you referring to the surprise of finding that Std.is would return false for (a, b) when a is null, even though cast(a, b) in the same situation would succeed?  In that case, I do agree that it is surprising, which is why I believe that modifying Std.is to have semantics matching cast is the right solution.  I only proposed Std.canCast as an alternative to this more correct solution, because ncannasse seems to not be convinceable that Std.is semantics should change.

**Comment by waneck** [comment]
Well, if anything I agree with Simn. The api of canCast / is difference is
weird. I don't think Std.is (null,SomeType) should ever return null. But I
agree this should be changed only in the next haxe version

**Comment by ncannasse** [comment]
@waneck you meant `cast`, right ? (Std.is will always return true/false) And by "version", please specify either minor OR major :) things gets more complicated if we are not sure who thinks what.

As for Std.is VS cast : of course it would be strange if something that "is" T cannot be cast to it, but that's not the case in what I'm proposing. So if we specify that cast is null-inclusive while Std.is is null-exclusive, then I there is no problem logical-wise. It might be a bit surprising but (a) that's already the case of most of our platforms and (b) it's not broken.

So I keep on proposing that we at least unify the behavior of the different platforms for next minor release (3.2) and think again about both "is" (this time as an operator?) and "cast" when we review their behavior wrt abstract support in 4.0

**Comment by bji** [comment]
@waneck, yes please clarify this statement:

"I don't think Std.is (null,SomeType) should ever return null"

Do you mean: "I don't think cast(null, SomeType) should ever return null"

Or do you mean: "I don't think Std.is(null, SomeType) should ever return false"

The difference is pretty much the difference between the two sides of this discussion, so clarification is kind of important :)

**Comment by Simn** [comment]
@bji: "not understanding what null means" refers to https://github.com/HaxeFoundation/haxe/issues/3848#issuecomment-73788537 because I simply don't understand how this makes any kind of sense.

You're right about what I meant by surprise-by-design. I could only accept this if we label Std.is as "not working as intended" with regards to null, but that hasn't really been on the table. If you guys say that we fix cast now and Std.is later then it's fine with me, although the introduced intra-target inconsistency still leaves a bad taste in my mouth and I feel like we're rushing things for no real reason here.

**Comment by bji** [comment]
Who's rushing?  Nobody has proposed changing anything except on a major release boundary.  I feel like what you mean by "rushing" is actually "taking the issue seriously", and why shouldn't we do that?

Also I think my post 3848 was very clear, so I don't know what could have confused you.  If you can clarify, I can try to address the confusion.

Finally, I think intra-target inconsistency is much preferable to inter-target inconsistency, because then I can write code once to a known set of (perhaps counterintuitive, but hopefully well documented and understandable) APIs, and not have to worry about whether or not some differing semantics on some other target is going to make my life hell when I decide to switch over to targeting a new platform.

Intra-target inconsistency simply requires me to become more well informed about haxe and its APIs.  Inter-target inconsistency requires me to waste time fixing stuff that shouldn't even have broken.

**Comment by Simn** [comment]
By rushing I mean that Nicolas wants to change this for 3.2 which I kind of want to release soonish.

The confusion comment wasn't yours, I linked to Nicolas comment which says "@Simn I don't see any issue here, that seems somewhat to match the definition of null.".

Regarding intra- vs inter-target inconsistency: I can see where you're coming from here and would probably prefer the latter as well for a new API (if I had to choose). I'm still not convinced that it's worth shaking up the the status quo to fix an inter-target inconsistency by introducing intra-target inconsistencies, especially if we can just do it cleanly and get rid of _both_ inconsistencies in the next major release (Haxe 4).

**Comment by bji** [comment]
Simn, I'd be onboard for your last statement, which would be to fix it all in a consistent manner for Haxe 4.  Can you describe briefly what that solution is?  If it's compelling I'd happily defer this particular issue while waiting for that.

**Comment by Simn** [comment]
* deprecate Std.is
* introduce `is` operator (#2976)
* make `null is T` yield `true` for any T which is not a basic type
* make `cast(null, T)` always succeed (not sure about basic types here)

In fact I would like to go a bit further and make a clear distinction between static and dynamic casts like in C++. I've seen _a lot_ of `cast(v, T)` usage which suggests that the user thinks it's a static cast while it's actually a dynamic cast with a built-in type-check.

**Comment by bji** [comment]
That all sounds great to me, except maybe the part about making a distinction between static and dynamic casts, which I find really painful in C++ (maybe it's just the syntax they use), but I wouldn't even argue against that if it's what you/others wanted to do.

**Comment by waneck** [comment]
I'm sorry. I meant that `Std.is(null,SomeTime)` should never return false. And I meant that I think this should be solved together with the `is` operator, presumably in the next major release only.

I feel it will be strange that `Std.is` and `cast` disagree about what can and cannot be cast, and in this sense I agree with Simn on this.

**Comment by Simn** [comment]
There is another player here in the form of try/catch:

```
class Main {
	static function main() {
		var n = null;
		var nMain:Main = cast(n, Main);
		// No exception so we know the cast was fine. Surely that means we can
		// catch it?
		try {
			throw nMain;
		} catch(e:Main) {
			trace("Caught " +e);
			Sys.exit(0);
		}
		// nope
		Sys.exit(1);
	}
}
```

This currently crashes on C++ because the cast works but the catch does not. The only way to catch a null value is by using Dynamic.

Surely this can't be the intended behavior?

**Comment by ncannasse** [comment]
Let's move forward with this.

We should implement the following in Haxe 3.x:

- cast(null, T) is always allowed (given that T is a Class/Enum)

Std.is and "catch" share the same behavior:
- Std.is(null,T) is always false
- null can only be caught by Dynamic

- cast(null,Int|Float|Bool) : not sure yet, it might depend on static platforms or not, and we need to decide on cast(null, Null...)

I'm fine after that to decide any change for 4.x or new "is" operator, but we need to fix 3.x behavior first

**Comment by waneck** [comment]
I agree with @Simn that it's strange that `cast` and `is` have different behaviours. Also, `is` for the `null` case will incur in a performance penalty for most targets, which return true for this case.
Also, another case for `Std.is(null, T)` to return true is that I think we should be consistent with our own type system. Since `null` is a valid `T` type, I see absolutely no reason for it to not return true in this case.

**Comment by ncannasse** [comment]
I don't agree it's strange to have a different behavior between cast and is.

Given that both operates at runtime, let's assume we are operating on a Dynamic value (which type is not known).

- `cast` means : Can I convert this runtime value to the given type ? This is true for null

- `is` means : Is this runtime value of the given type ? In that case the answer is different for null, since the null does not have any actual runtime type.

"Being" of the type and "being able to cast" to the type are different things IMHO.

Also, the two operators are used is quite different manners. Usually you want to do this kind of thing:

`if( Std.is(v,String) ) return escapeString(v);`

`var v = cast(v,MyClass); v.myMethod();`

I still push for proposal of unifying the platforms behavior as I proposed.

However I could live with cast(null,T) throwing an exception, but I definitely don't want to have Std.is(null,T) returning true.

**Comment by mcheshkov** [comment]
What about `Std.is(v, Null<String>)` via typedef ? `null` certainly is value of that type.

**Comment by ncannasse** [comment]
@mcheshkov from my point of view, null as a runtime value does not have a type, it does not carry any type information. Of course, telling it is of all types works as well, but `Null<T>` won't change anything here since String is nullable.

**Comment by Simn** [comment]
Reading through everything again, I stand by my opinion from 2 years ago. In particular I'm calling mental gymnastics on this point of view:

> cast means : Can I convert this runtime value to the given type ? This is true for null
>
> is means : Is this runtime value of the given type ? In that case the answer is different for null, since the null does not have any actual runtime type.
>
> "Being" of the type and "being able to cast" to the type are different things IMHO.

You say that cast converts a run-time value to a given type. Surely that means that the result _is of the given type_. But that wouldn't hold for null: `(cast (null, T)) is T` would yield false. This is not logically sound.

It's also not a good basis for design to rely on terminological nuances like that. This just leads to surprises...

**Comment by back2dos** [comment]
I fully agree with Simn. What would be the point anyway? It's not like `cast (v, Foo).methodOfFoo()` will do anything meaningful if `v == null`. 

Unless you're reasonably sure that the value is not `null`, you will have to check it at some point anyway. You might as well do it before the cast, rather than after. 

Raising an exception when casting `null` gives stronger guarantees and makes errors surface earlier, both of which are Good Things™.

**Comment by ncannasse** [comment]
I read it again:

`cast(null,T)` should be allowed, and have the same result as `var x : T = cast null`.
like it or not, null is a valid value for T, `cast` does not check if the value is "strictly of type T" (since T subclasses are allowed), but if the they are "values that can be cast to T"

In case you want to allow null, that would mean writing stupid things such as:  `if( v == null ) null else cast(v,T)` whereas you can always check for null later. null is not an error value, it's a correct value and should be treated as it.

If you really insist that Std.is to be changed to match the specification of cast, then I can go with it although I feel this is breaking a lot of code for the sake of NO actual win other than some abstract definition of having things done "right" -- I'm currently perfectly fine with having Std.is slightly different in definition than cast.

**Comment by kevinresol** [comment]
> "Being" of the type and "being able to cast" to the type are different things IMHO.

IMHO: this statement only holds if the conversion done by `cast` could involve a transformation process (i.e. creating a new object). Otherwise if it is always returning the same object/reference, then logically "being able to cast" === "being of the type".

**Comment by back2dos** [comment]
> I feel this is breaking a lot of code for the sake of NO actual win other than some abstract definition of having things done "right" 

If you could come forward with something more tangible than your gut feeling, that'd be great. 

As it is, the cast throws an error for js, flash and neko and quite possibly other targets (haven't checked). But as far as I can tell, the vast majority of all libraries are tested on at least one of those targets, so it's safe to assume they will be unaffected.

And again, I just don't see the use case. Let's assume there is some code in production somewhere, that performs the cast on `null` and relies on it to not throw. It must then be either of these:

```haxe
var t = cast (v, T);
//do nothing with t .... ever
```

or:

```haxe
var t = cast (v, T);
if (t != null) {
  //do something with t
}
```

The first one is outright non-sense. The second one is indeed not entirely unrealistic, but I think it's fair to assume that the vast majority of people will put the null-guard first, because it's the cheaper check to perform.

So my argument: Having the cast not throw doesn't have any practical benefit, while having it throw does. I don't know if that makes it "right", but suppose it does, do you genuinely believe that so many people rely on this behavior that we should not fix it?

**Comment by hughsando** [comment]
For me, starting from the top,
I like the idea that cast(x,T) is a "safe cast" - will never return null - that is what it means to be safe.  So even though null can be assigned to x of type T, it is defeating the entire purpose of "safe" cast.

Std.is - I'm not sure there is a 100% logical answer, because you can always tweak your definitions to suit the outcome. So the two alternatives:
a) Std.is(null,T) = true
b) Std.is(null,T) = false
Which one of these will make it easier for developers to write correct code?
The main use cases I can think of are 1, a multi-type:
```haxe
   if (/*x==null ||*/ Std.is(x,T1)) runT1(cast x);
   else if (Std.is(x,T2)) runT2(cast x);
   else throw "Unexpected type";
```
And a 2 pre-condition safety check/ "return code fail":
```haxe
  if (Std.is(x,T1)) { runT1(cast x); return true; }
  return false;
```
Again, I don't think either case if definitive, but it is(null) returns false, then it requires the developer to explicitly handle the case in the multi-type example, which I think might be better.  For the pre-condition  case I would needs some kind of survey of existing code to be convinced either way.

So my preference would be to say that null is not actually any time,  that is "casting will return a non-null value"  and safe-cast always returns non-null instances of the matching type, or throws.

**Comment by Simn** [comment]
To clarify my viewpoint: I strongly believe that Std.is and safe-casts should be consistent. Whether or not this means allowing null-values is something I'm not so sure about. It's probably a theory (allow null) vs. pragmatism (don't allow null) thing, and we often side with pragmatism on these...

**Comment by hughsando** [comment]
I think we are on the same page.  I agree they should be consistent, and I think it is most practical to throw on null.

**Comment by Simn** [comment]
Oh god this issue...

**Comment by jcward** [comment]
Pragmatically :wink:  I agree with Hugh and Nicolas.

Basically, `Std.is` is a core function that should make users feel safe. At runtime (and Std.is is a runtime concern), "safety" implies not only type safety but null safety, so the user expects:
- the value is a valid Foo and I can operate on it.
- the value can be cast to Foo and I expect to be able to use the result (non-null).
- Std.is can be used in an if/else-if priority structure, so the first clause to hit is the valid type.

Thus, to fulfill those expectations, `Std.is(null)` returns false.

Assuming the burden of null checking is unavoidable, please place it on the coder who's depending on casting null, instead of the coder who's trying to be a good citizen using `Std.is`.

***Aside:*** If there's a performance penalty, then add `Std.fastIs` with a note about null (I suppose that's the difficulty with an `is` operator. It's got to be one or the other...)
